File: js/trees/0103-binary-tree-zigzag-level-order-traversal.js
var zigzagLevelOrder = function(root) {
    var response = [];
    if (root === undefined) return response;
    
    var queue = [];
    queue.push(root);
    
    while (queue.length > 0) {
        var level = []
        var len = queue.length;
    
        for (var i  = 0; i < len; i++) {
            var cur = queue.shift();
            
            level.push(cur.val);

            if (cur.left) queue.push(cur.left);
            if (cur.right) queue.push(cur.right);
        }
        response.push(level);
    }
    
    for (var i = 0; i < response.length; i++) {
        if (i % 2 !== 0) {
            response[i].reverse();
        }
    }
    
    return response;
};

File: cpp/trees/0103-binary-tree-zigzag-level-order-traversal.cpp
#include <algorithm>
#include <queue>
#include <vector>
#include <structures.h>

using namespace std;


vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
    vector<vector<int>> response;
    if (root == nullptr) return response;
    
    queue<TreeNode*> q;
    q.push(root);
    bool mustReverse = false;
    
    while (!q.empty()) {
        vector<int> level;
        
        int len = q.size();
        
        for (int i = 0; i < len; i++) {
            TreeNode* cur = q.front();
            q.pop();
            
            level.push_back(cur->val);
            
            if (cur->left) q.push(cur->left);
            if (cur->right) q.push(cur->right);
        }
        
        if (mustReverse) {
            reverse(level.begin(), level.end());
        }
        
        mustReverse = !mustReverse;
        response.push_back(level);            
    }
    
    return response;
}